### PR TITLE
Better arguments

### PR DESCRIPTION
**Pull request by benwilson512**
Solves issues with heavily nested arguments and variables.

### CODE REVIEW

**Comment by benwilson512** [comment]
TODO:
- [ ] Verify if the built in directives should have non null inputs w/o defaults
- [ ] Add deprecation checking
- [ ] Add missing checking back to arguments
- [ ] Figure out the best route for returning `{:error, values}` or not from an execution run.

**Comment by benwilson512** [comment]
On that last note @bruce it looks like Absinthe.run is supposed to return `{:ok, stuff}` for nearly all cases, which is fine.

**Comment by bruce** [comment]
RE directives: looking at the [spec](https://facebook.github.io/graphql/#sec-Type-System.Directives), I'd say the `if` argument for both `@skip` and `@include` should be `non_null(:boolean)`

**Comment by bruce** [comment]
RE errors -- correct -- we almost always want to push errors to the API user via the [errors property of the response](https://facebook.github.io/graphql/#sec-Errors).

**Comment by benwilson512** [comment]
This test case https://github.com/absinthe-graphql/absinthe/blob/master/test/lib/absinthe/type/directive_test.exs#L91 would seem to indicate that for these boolean directives not including them should default to false.

File: lib/absinthe/type/built_ins/scalars.ex [hunk 1]
```diff
@@ -102,13 +102,10 @@ defmodule Absinthe.Type.BuiltIns.Scalars do
...
-  defp parse_string(value) do
-    try do
-      {:ok, to_string(value)}
-    rescue
-      Protocol.UndefinedError -> :error
-    end
+  defp parse_string(value) when is_binary(value) do
+    {:ok, value}
   end
+  defp parse_string(_), do: :error
```
Comment (benwilson512) [inline lib/absinthe/type/built_ins/scalars.ex]: @bruce I think this is a better route for parsing strings. `Kernel.to_string` is entirely too forgiving with inputs. For example it'll take a list of integers `[1,2,3]` and turn it into `<<1,2,3>>` which is definitely not gonna be considered normal.

I could see an argument for coercing floats / integers to binaries, but I think that ought to be the limit.

**Pull request merged**